hoomd/data/data_structures.py

```diff
@@ -0,0 +1,497 @@
+    consistency with C++ after attaching.
+    """
+    @abstractmethod
+    def to_base(self):
+        """Cast the object into the corresponding native Python container type.
+        """
+        pass
+
+    @contextmanager
+    def _buffer(self):
+        """Allows buffering during modifications to prevent needless updates.
+
+        This is a simple system to prevent unnecessary C++ - Python
+        communication.
+        """
+        self._buffered = True
+        yield None
+        self._buffered = False
+
+    def _update(self):
+        """Signal that the object has been updated."""
+        if self._buffered:
+            return
+        elif parent is not None:
+            self._parent._handle_update(self, self._label)
+
+
+def _get_inner_typeconverter(type_def, desired_type):
+    """Handles more complicated data structure specification.
+
+    When using nested containers where an entry could be a list or None for
+    instance requires the use of something like typeconverter's OnlyIf or
+    Either. We autdetect the internal type specification for these object and if
+    they are of the correct type use these.
+
+    Args:
+        type_def (`hoomd.data.typeconverter.TypeConverter`):
+            The type converter which is desired to be introspected to find the
+            desired type.
+        desired_type (`hoomd.data.typeconverter.TypeConverter`):
+            A child class of `hoomd.data.typeconverter.TypeConverter` that
+            signals what type converter types is needed.
+    """
+    rtn_type_def = None
+    # Simple case the type_def is already sufficient
+    if isinstance(type_def, desired_type):
+        rtn_type_def = type_def
+    # More complicated cases, deals with nested type definitions where the
+    # nested containers have other requirements such as multiple valid container
+    # representations or entry is allowed to be None. In these cases, we check
+    # the TypeConverterValue instance's converter for the desired container type
+    # converter type.
+    elif isinstance(type_def, TypeConverterValue):
+        if isinstance(type_def.converter, OnlyIf):
+            if isinstance(type_def.converter.cond, desired_type):
+                rtn_type_def = type_def.converter.cond
+        elif isinstance(type_def.converter, Either):
+            for spec in type_def.converter.specs:
+                matches = []
+                if isinstance(spec, desired_type):
+                    matches.append(spec)
+                if len(matches) > 1:
+                    raise ValueError(
+                        "Parameter defined with multiple valid definitions of "
+                        "the same type.")
+                elif len(matches) == 1:
+                    rtn_type_def = matches[0]
+    # When no valid type converter can be found error out.
+    if rtn_type_def is None:
+        raise ValueError("Unable to find type definition for attribute.")
+    return rtn_type_def
```

Comment:
I would try to separate out the comments so that "multiple valid container representations" or "allowed to be None" align with the checks where those are implemented. Part of the challenge is that I don't know what the `OnlyIf` or `Either` classes are meant to do. Reading their docstrings didn't help, the `OnlyIf` docstring mentions "Not strictly necessary, but keeps the theme of the other classes, and allows pre/post-processing and optionally allows None." Is that class meant to act like [typing.Optional](https://docs.python.org/3/library/typing.html#typing.Optional)?
```suggestion
    # More complicated cases, deals with nested type definitions where the
    # nested containers have other requirements such as multiple valid container
    # representations or entry is allowed to be None. In these cases, we check
    # the TypeConverterValue instance's converter for the desired container type
    # converter type.
    return_type_def = None
    if isinstance(type_def, TypeConverterValue):
        if isinstance(type_def.converter, OnlyIf):
            if isinstance(type_def.converter.cond, desired_type):
                return_type_def = type_def.converter.cond  # Can this be None? If not, return here.
        elif isinstance(type_def.converter, Either):
            matches = [spec for spec in type_def.converter.specs if isinstance(spec, desired_type)]
            if len(matches) > 1:
                raise ValueError(
                    "Parameter defined with multiple valid definitions of "
                    "the same type.")
            elif len(matches) == 1:
                return_type_def = matches[0]
    # When no valid type converter can be found error out.
    if return_type_def is None:
        raise ValueError("Unable to find type definition for attribute.")
    return return_type_def
```